src/schematools/types.py

```diff
@@ -24,11 +37,11 @@ def __missing__(self, key: str) -> NoReturn:
 
     @property
     def id(self) -> str:
-        return self["id"]
+        return str(self["id"])
 
     @property
     def type(self) -> str:
-        return self["type"]
+        return str(self["type"])
```

Comment:
misschien liever `cast(str, self["type"])`?

Reply:
The conversions or casts are necessary as MyPy has no idea what the types of the values associated with the keys `id` and `type` are. But I'm wondering if we can do better?

Schemas have a set a predefined attributes and a set of user definable attributes. If we would have had only predefined attributes we could have used a `dataclass` (or `TypedDict`). That way MyPy would have know that both `id` and `type` are of type`str`, obviating the need for casts/conversions.

However both `dataclass`es and `TypedDict`s do not support extra keys (the user definable attributes in the schemas, eg column names of tables). Hence we can't use a `dataclass` or even a `TypedDict` as the base class for our schema types. I'm wondering if there is anything out there that would allow for extra keys in Python data structures while still providing the type checking for the predefined keys?

Reply:
Pydantic seems to be able to provide the means to do exactly that with the following option:
```python
    class Config:
        extra = 'allow'
```

See also: https://pydantic-docs.helpmanual.io/usage/model_config/

* extra
  *  whether to ignore, allow, or forbid extra attributes during model initialization. Accepts the string values of 'ignore', 'allow', or 'forbid', or values of the Extra enum (default: Extra.ignore). 'forbid' will cause validation to fail if extra attributes are included, 'ignore' will silently ignore any extra attributes, and 'allow' will assign the attributes to the model.